DOCS: lstk migration guide (#873)
localstack/localstack-docs: src/content/docs/aws/developer-tools/running-localstack/lstk-migration.mdx

---
title: lstk Migration Guide 
description: Migration guide for lstk, the new command-line interface for LocalStack.
template: doc
sidebar:
  order: 11
tags: ['Hobby']
---

# Migrating from `localstack` to `lstk`

## Introduction

`lstk` is the new command-line interface for LocalStack. It is a single, self-contained binary that starts and manages the emulator, runs the AWS CLI and your infrastructure-as-code tools, and saves and restores emulator state. In other words, it replaces both the Python-based `localstack` CLI and the family of wrapper scripts that many projects installed alongside it. Those scripts are `awslocal`, `tflocal`, `samlocal`, and `cdklocal`.

> There are three good reasons to start using `lstk`. The first is that installation is simpler, because there is no Python environment to manage, and no separate wrapper script to install for each tool you use. The second is that `lstk` works with every LocalStack emulator, not just AWS. Snowflake and Azure are supported today, and future emulators will be available too. The third is that `lstk` is the place for new CLI functionality to be added from now on. The `localstack` CLI is deprecated, and will no longer be supported.

Reviewer: I think this formatting makes it easier to read/skim

```suggestion
There are three good reasons to start using `lstk`:

1. Installation is simpler, because there is no Python environment to manage, and no separate wrapper script to install for each tool you use.
2. `lstk` works with every LocalStack emulator, not just AWS. Snowflake and Azure are supported today, and future emulators will be available too.
3. `lstk` is the place for new CLI functionality to be added from now on. The `localstack` CLI is deprecated, and will no longer be supported.
```

What does not change is LocalStack itself. The emulator is the same Docker image with the same behavior and the same configuration variables (`DEBUG`, `SERVICES`, `PERSISTENCE`, and the rest). You are changing the tool you drive LocalStack with, not the LocalStack emulator itself. For most teams the migration is a short exercise in translating a handful of commands in shell history, scripts, and CI pipelines.

:::note 
The legacy CLI and the wrapper scripts continue to work, so you can migrate gradually rather than all at once. A small number of features are still only available with the legacy `localstack` CLI. Those are listed under [Unsupported features](#unsupported-features).
:::

## Installation

`lstk` is distributed through the LocalStack Homebrew tap and the `@localstack/lstk` npm package. Pre-built binaries for Linux, macOS, and Windows are published on GitHub Releases. Pick whichever fits your environment.

> With Homebrew:

Reviewer: Would this look better as a code block with two tabs labeled Homebrew and npm?

```bash
brew install localstack/tap/lstk
```

With npm:

```bash
npm install -g @localstack/lstk
```

> Check for correct installation by invoking `lstk --version`. Docker must be installed and running, exactly as before. Homebrew installs shell completions automatically, although with the other methods you can generate them yourself, using `lstk completion bash|zsh|fish|powershell`. Full instructions are in the lstk documentation.

Reviewer: ```suggestion
Check for correct installation by invoking `lstk --version`. Docker must be installed and running, exactly as before. Homebrew installs shell completions automatically. With the other methods you'll need generate them yourself, using `lstk completion bash|zsh|fish|powershell`. Full instructions are in the [lstk documentation](https://docs.localstack.cloud/aws/developer-tools/running-localstack/lstk/).
```

There is no need to uninstall the `localstack` CLI, as the two can exist side by side. This is useful while you migrate, and necessary if you rely on a feature that is not yet available with `lstk`. However, we recommend against using them at the same time. Each CLI starts and manages its own LocalStack container, so stop whichever is running before you start the other.

## Logging in

Both CLIs need a LocalStack license to run the emulator, but they ask for it in different ways. With the `localstack` CLI you copied an auth token out of the web application and stored it with `localstack auth set-token`, or exported `LOCALSTACK_AUTH_TOKEN` yourself.

`lstk` replaces that with a browser-based login. Run `lstk login` and approve the request in the browser window that opens. The credential is then stored securely on your machine. There is no token to copy, and none to keep in your shell profile.

```bash
lstk login
```

Similarly, use `lstk logout` to remove the auth token from your machine.

> Continuous integration has no browser, so auth tokens remain the right approach there. Set `LOCALSTACK_AUTH_TOKEN` as a secret in your pipeline, and `lstk` will use it without any login step. If necessary, use a CI Auth Token rather than a personal developer token, as described in the Auth Token documentation.

Reviewer: CI tokens are definitely what we want them to use.

```suggestion
Continuous integration has no browser, so auth tokens remain the right approach there. Set `LOCALSTACK_AUTH_TOKEN` as a secret in your pipeline, and `lstk` will use it without any login step. Use a CI Auth Token rather than a personal developer token, as described in the [Auth Token documentation](https://docs.localstack.cloud/aws/getting-started/auth-token/#ci-environments).
```

    ```
  </TabItem>
  <TabItem label="npm">
    ```bash
    npm install -g @localstack/lstk
    ```
  </TabItem>
</Tabs>

Check for correct installation by invoking `lstk --version`. Docker must be installed and running, exactly as before. Homebrew installs shell completions automatically. With the other methods you'll need generate them yourself, using `lstk completion bash|zsh|fish|powershell`. Full instructions are in the [lstk documentation](https://docs.localstack.cloud/aws/developer-tools/running-localstack/lstk/).

There is no need to uninstall the `localstack` CLI, as the two can exist side by side. This is useful while you migrate, and necessary if you rely on a feature that is not yet available with `lstk`. However, we recommend against using them at the same time. Each CLI starts and manages its own LocalStack container, so stop whichever is running before you start the other.

## Logging in

Both CLIs need a LocalStack license to run the emulator, but they ask for it in different ways. With the `localstack` CLI you copied an auth token out of the web application and stored it with `localstack auth set-token`, or exported `LOCALSTACK_AUTH_TOKEN` yourself.

> `lstk` replaces that with a browser-based login. Run `lstk login` and approve the request in the browser window that opens. The credential is then stored securely on your machine. There is no token to copy, and none to keep in your shell profile.

Reviewer 2: I'd mention simply running `lstk` or `lstk start` will take care of the auth flow, then say run `lstk login` if the user specifically wants to authenticate or they want to switch account.

| `localstack start --host-dns` | `expose_ports = [53]` |
| `AUTO_LOAD_POD=my-baseline` | `snapshot = "pod:my-baseline"` |

To illustrate, the following before-and-after shows the mapping. Where you previously ran:

```bash
DEBUG=1 PERSISTENCE=1 localstack start -d \
  -e SERVICES=s3,sqs \
  -v ./init.sh:/etc/localstack/init/ready.d/init.sh
localstack wait
```

you would now write a `.lstk/config.toml` in the project.

```toml
[[containers]]
type    = "aws"
port    = "4566"
env     = ["dev"]
volumes = ["./init.sh:/etc/localstack/init/ready.d/init.sh"]

[env.dev]
DEBUG = "1"
SERVICES = "s3,sqs"
PERSISTENCE = "1"
```

then start it with `lstk start`. Because the `config.toml` file lives in the repository, everyone on the team gets the same environment. Your CI jobs also pick up the same file.

The existing `CONFIG_PROFILE` mechanism in the `localstack` CLI and its `~/.localstack/*.env` files have no direct equivalent with `lstk`. The recommended replacement is a per-project `.lstk/config.toml`. If you need to switch between several configurations for the same project, keep them as separate files and choose between them with `lstk --config <path>`.

> ## Unsupported features

Reviewer 2: I'd link the feature name in the table to the feature docs page so it's easier to find, and also since we mention here "you should follow the linked documentation for more detail"

| Cloud Pod remotes other than S3 | Use the `localstack` CLI. `lstk` supports the LocalStack platform and your own S3 bucket, with the location passed inline rather than registered in advance. |
| Host DNS setup on Linux | Use the `localstack` CLI for the one-off resolver configuration. Publishing the DNS port itself is supported through the `expose_ports` setting. |
| Arbitrary Docker options, such as custom networks | Use `docker-compose.yml` when you need full control of the container and the many Docker configuration options. The `lstk` configuration file covers only the basic features, such as the image, tag, port, exposed ports, volumes, and environment variables. |
| Shell access to the container (`localstack ssh`) | Use `docker exec -it localstack-aws bash`. Note that `lstk` names its container `localstack-[aws,azure,snowflake]`. |
| GitHub Action | There is no GitHub Action for `lstk` yet, and the existing setup-localstack Action installs the `localstack` CLI. Install `lstk` in a step of your own instead, as shown in [Using lstk with Continuous Integration](#using-lstk-with-continuous-integration). |

> ## Next steps 

Reviewer 2: Similarly, it'll be helpful to link the relevant docs for easier navigation.
